KAFKA-16445: Add PATCH method for connector config

ivanyu opened the pull request:
This is the implementation of [KIP-477: Add PATCH method for connector config in Connect REST API](https://cwiki.apache.org/confluence/display/KAFKA/KIP-477%3A+Add+PATCH+method+for+connector+config+in+Connect+REST+API). This PR introduces the new endpoint in the Connect REST API which allows patching the configuration of an existing connector. The patch is applied on top of the config of the existing connector. `null` values work as tombstones, i.e. make the key disappear.

The change is tested on the unit and integration levels.

ivanyu committed:
KAFKA-16445: Add PATCH method for connector config

This is the implementation of https://cwiki.apache.org/confluence/display/KAFKA/KIP-477%3A+Add+PATCH+method+for+connector+config+in+Connect+REST+API

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -992,4 +992,15 @@ public List<String> setWorkerLoggerLevel(String namespace, String desiredLevelSt
+            }
+        });
+        return patchedConfig;
+    }
```
This feels generic enough that we could probably put it into [ConnectUtils](https://github.com/apache/kafka/blob/e63efbc5d771a6e69eb8678db6f88730b4e751e1/connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java) or even just [Utils](https://github.com/apache/kafka/blob/e63efbc5d771a6e69eb8678db6f88730b4e751e1/examples/src/main/java/kafka/examples/Utils.java) as something like `patchConfig` (if in the former) or `combineMaps`, `mergeMaps`, or `patchMap` (if in the latter).

ivanyu replied:
I did this, extracted into `ConnectUtils` (because I think the `null` processing logic may be connect-specific) + added a unit test for this.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -242,6 +242,19 @@ public Response putConnectorConfig(final @PathParam("connector") String connecto
+        herder.patchConnectorConfig(connector, connectorConfigPatch, cb);
+        Herder.Created<ConnectorInfo> createdInfo = requestHandler.completeOrForwardRequest(cb, "/connectors/" + connector + "/config",
+                "PATCH", headers, connectorConfigPatch, new TypeReference<ConnectorInfo>() { }, new CreatedConnectorInfoTranslator(), forward);
+        return Response.ok().entity(createdInfo.result()).build();
```
Just realizing now that we don't actually specify the status and body of the REST response in the KIP. I agree with what's here, especially since it matches  the existing `PUT /connectors/{name}/config` endpoint, but it's worth specifying in the KIP for completeness.

ivanyu replied:
I didn't find the current `PUT` response documentation, so I propose something like this:
```
Responses follow the model of the configuration PUT endpoint:
1. If the patch was successfully applied, the response code is 200 and the body is a JSON object with the updated connector information (`name`, `type`, `config`, and `tasks`), for example:
{
    "name": "my-connector",
    "config":
    {
        "name": "my-connector",
        "sample_config_2": "test_config_2",
        "sample_config": "test_config_new"
    },
    "tasks":
    [
        {
            "connector": "my-connector",
            "task": 0
        },
        {
            "connector": "my-connector",
            "task": 1
        }
    ],
    "type": "sink"
}

2. In case of errors, the response code matches the error type (e.g. 400 in case of a config validation error; 404 if the connector is not found; 500 in case of other server-side errors) and the body is a JSON object with the error details:
{
    "error_code": 400,
    "message": "Connector configuration is invalid and contains the following 1 error(s):\n...\nYou can also find the above list of errors at the endpoint `/{connectorType}/config/validate`"
}
```
WDYT?

C0urante replied:
Yep, LGTM 👍

Can you add that to the KIP?

ivanyu replied:
Yes, done

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1144,6 +1144,21 @@ public void putConnectorConfig(final String connName, final Map<String, String>
+    @Override
+    public void patchConnectorConfig(String connName, Map<String, String> configPatch, Callback<Created<ConnectorInfo>> callback) {
+        log.trace("Submitting connector config patch request {}", connName);
+        addRequest(() -> {
```
Can we do a check to make sure we're the leader at the beginning of the request body here? This would reduce (but not completely eliminate) the chance for race conditions.

ivanyu replied:
I added this check + a test for this (does it make sense to have it?)

C0urante replied:
Yep, looks great!

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -242,6 +242,19 @@ public Response putConnectorConfig(final @PathParam("connector") String connecto
+    @Path("/{connector}/config")
+    public Response patchConnectorConfig(final @PathParam("connector") String connector,
+                                         final @Context HttpHeaders headers,
+                                         final @QueryParam("forward") Boolean forward,
```
Nit:
```suggestion
                                         final @Parameter(hidden = true) @QueryParam("forward") Boolean forward,
```

ivanyu replied:
Thanks, fixed

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -246,6 +246,31 @@ private synchronized void putConnectorConfig(String connName,
     }
 
+    @Override
+    public void patchConnectorConfig(String connName, Map<String, String> configPatch, Callback<Created<ConnectorInfo>> callback) {
```
Should we make this method `synchronized` in order to prevent (or at least reduce) race conditions?

ivanyu replied:
Yes, makes sense. Done

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -246,6 +246,31 @@ private synchronized void putConnectorConfig(String connName,
+                        () -> putConnectorConfig(connName, patchedConfig, null, true, callback, configInfos)
+                );
+            });
+        } catch (ConnectException e) {
```
I don't know why we limit the catch clause to `ConnectException` in `deleteConnectorConfig`, but here (and probably there as well) it should be `Throwable` instead; otherwise, the callback is left dangling.

ivanyu replied:
Ack, fixed

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -2336,6 +2336,95 @@ public void testPutConnectorConfig() throws Exception {
+
+        Map<String, String> originalConnConfig = new HashMap<>(CONN1_CONFIG);
+        originalConnConfig.put("foo1", "bar1");
+        originalConnConfig.put("foo2", "bar2");
```
Nit: can we add one more key/value pair that should be unchanged after the patch is applied? This would catch bugs where the set of post-patch keys is derived from the patch instead of the combination of the patch and the prior configuration.

ivanyu replied:
Yes, good idea. Done here and in `StandaloneHerderTest` too.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1144,6 +1144,21 @@ public void putConnectorConfig(final String connName, final Map<String, String>
+                callback.onCompletion(new NotFoundException("Connector " + connName + " not found", null), null);
+            } else {
+                Map<String, String> patchedConfig = applyConnectorConfigPatch(connectorInfo.config(), configPatch);
+                putConnectorConfig(connName, patchedConfig, true, callback);
```
This adds another request to the queue instead of executing the follow-up logic immediately, which increases the chance for race conditions. Can we fix that? We could possibly have a synchronous variant of `putConnectorConfig` whose method body is pulled from [here](https://github.com/apache/kafka/blob/e63efbc5d771a6e69eb8678db6f88730b4e751e1/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L1099-L1140), and invoke that variant instead of the asynchronous one at this line in `patchConnectorConfig`.

ivanyu replied:
Yes, makes sense, done.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -2336,6 +2336,95 @@ public void testPutConnectorConfig() throws Exception {
+            return null;
+        }).when(herder).validateConnectorConfig(eq(patchedConnConfig), validateCallback.capture());
+
+        // This is effectively the main check of this test:
+        // we validate that what's written in the config storage is the patched config.
+        doNothing().when(configBackingStore).putConnectorConfig(eq(CONN1), eq(patchedConnConfig), isNull());
```
Nit: could we change this to an explicit `verify` invocation?

```suggestion
        // This is effectively the main check of this test:
        // we validate that what's written in the config storage is the patched config.
        verify(configBackingStore).putConnectorConfig(eq(CONN1), eq(patchedConnConfig), isNull());
```

And, if we want to be extra fancy, we can make sure that no other configs were written:
```java
        verifyNoMoreInteractions(configBackingStore);
```

ivanyu replied:
Yes, makes sense, applied

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -2336,6 +2336,95 @@ public void testPutConnectorConfig() throws Exception {
+
+        expectMemberEnsureActive();
+        expectRebalance(1, Arrays.asList(CONN1), Collections.emptyList(), true);
+        when(worker.getPlugins()).thenReturn(plugins);
```
This part is unused, which causes the `DistributedHerder` test suite to fail when run as a whole thanks to `@RunWith(MockitoJUnitRunner.StrictStubs.class)`. We can and should remove it:
```suggestion
```

ivanyu replied:
Thanks, fixed

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java:
```diff
@@ -751,6 +754,74 @@ public void testPutConnectorConfig() throws Exception {
+        ArgumentCaptor<NotFoundException> exceptionCaptor = ArgumentCaptor.forClass(NotFoundException.class);
+        verify(patchCallback).onCompletion(exceptionCaptor.capture(), isNull());
+        assertEquals(exceptionCaptor.getValue().getMessage(), "Connector " + CONNECTOR_NAME + " not found");
+        assertNull(exceptionCaptor.getValue().getCause());
```
Just curious, what was the rationale for this assertion?

ivanyu replied:
I think, none. Removed

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java:
```diff
@@ -751,6 +754,74 @@ public void testPutConnectorConfig() throws Exception {
+        // Create the connector.
+        Map<String, String> originalConnConfig = connectorConfig(SourceSink.SOURCE);
+        originalConnConfig.put("foo1", "bar1");
+        originalConnConfig.put("foo2", "bar2");
```
Same suggestion RE adding another k/v pair that stays unchanged before/after the patch.

ivanyu replied:
Done

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java:
```diff
@@ -751,6 +754,74 @@ public void testPutConnectorConfig() throws Exception {
+        patchedConnConfig.put("foo3", "added");
+
+        expectAdd(SourceSink.SOURCE);
+        Connector connectorMock = mock(SourceConnector.class);
```
Unused variable?

ivanyu replied:
Yeah, a rebase artifact, deleted.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java:
```diff
@@ -751,6 +754,74 @@ public void testPutConnectorConfig() throws Exception {
+        assertEquals(patchedConnConfig, returnedConfig2);
+    }
+
+    private void expectConnectorStartingWithoutTasks(Map<String, String> config) {
```
Nit: just to avoid confusion if someone leverages this method in the future, could we either rename it to `expectSourceConnectorStartingWithoutTasks`, or add a `SourceSink connectorType` parameter to it that can be used to control which kind of connector is expected?

ivanyu replied:
Thanks, did the latter

C0urante reviewed (commented):
Thanks Ivan! Looks pretty good, have some suggestions but they should hopefully be on the simpler side to address.

One final request--can we add an integration test case for this new endpoint, probably to the [ConnectWorkerIntegrationTest suite](https://github.com/apache/kafka/blob/e63efbc5d771a6e69eb8678db6f88730b4e751e1/connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java)?

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Extract `patchConfig` into `ConnectUtils`

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Make `forward` hidden in `patchConnectorConfig`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Make `patchConnectorConfig` synchronized in `StandaloneHerder`

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Catch `Throwable` in `patchConnectorConfig`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Make connector config patching cover all cases of key/value changes in tests.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Better verification in `testPatchConnectorConfig`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Remove unused mock in `DistributedHerderTest`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Remove unnecessary assert in `StandaloneHerderTest`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Remove unused variable in `StandaloneHerderTest`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Make `expectConnectorStartingWithoutTasks` in `StandaloneHerderTest` applicable to both connector types.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Explicitly check we're the leader in `DistributedHerder`.

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Don't place an extra request to reduce the chance for races in `DistributedHerder`.

ivanyu committed:
fixup! fixup! KAFKA-16445: Add PATCH method for connector config

ivanyu committed:
fixup! KAFKA-16445: Add PATCH method for connector config

Add integration test

ivanyu commented:
I also added the integration test as you proposed @C0urante

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -2336,6 +2336,133 @@ public void testPutConnectorConfig() throws Exception {
+        connConfigPatch.put("foo2", null);
+        connConfigPatch.put("foo3", "added");
+
+        Map<String, String> patchedConnConfig = new HashMap<>(originalConnConfig);
```
Nit: we overwrite every value in this map, no need to clone the original
```suggestion
        Map<String, String> patchedConnConfig = new HashMap<>();
```

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java:
```diff
@@ -773,6 +773,41 @@ public void testDeleteConnectorCreatedWithPausedOrStoppedInitialState() throws E
+        expectedConfig.put("to-be-added-key", "value");
+        expectedConfig.put(TASKS_MAX_CONFIG, "2");
+        expectedConfig.remove("to-be-deleted-key");
+        assertEquals(expectedConfig, connect.connectorInfo(CONNECTOR_NAME).config());
```
I think it's possible for poor timing (which Jenkins is notorious for...) to create flakiness here. The connector and both of its tasks may be started, but it's possible that the worker we hit with this request won't have read the patched connector config from the config topic yet if it's not the leader of the cluster.

As a quick hack, we could tweak the order of operations and rely on existing retry logic in `ConnectAssertions::assertConnectorAndExactlyNumTasksAreRunning` to prevent this:

1. Configure connector with `tasks.max = 2`
2. Ensure connector is started and 2 tasks are running
3. Patch connector, including changing `tasks.max` to `3`
4. Ensure connector is started and 3 tasks are running
5. Perform the assertion on this line (i.e., that the connector config as reported by an arbitrary worker in the cluster matches the expected patch config)

ivanyu replied:
Thanks, great suggestion. Done

ivanyu committed:
fixup! fixup! KAFKA-16445: Add PATCH method for connector config

Reduce potential flakiness of the integration test

ivanyu commented:
Thank you @C0urante. I updated the description (please tell me if I need to add more) there.
Do you want to me squash the fixups or you squash when merge?

C0urante reviewed (approved):
Thanks @ivanyu, LGTM!

We squash merge in this project, so no need to do the squashing/force-pushing yourself. I'll wait for CI to complete and if everything looks alright, merge to trunk after.
